### PR TITLE
[profiler] Add  kineto init delay when used in daemon mode

### LINKED ISSUE #112389: KINETO_USE_DAEMON not work
# Description
When I set the env  `KINETO_USE_DAEMON=1`.
A cupti error `CUPTI_ERROR_MULTIPLE_SUBSCRIBERS_NOT_SUPPORTED` will be reported when I import torch in python interpreter.

# Shell Stdout

4029780fc904(@:):test# export KINETO_USE_DAEMON=1
4029780fc904(@:):test# python3

Python 3.9.2 (default, Feb 28 2021, 17:03:44)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
\>>> import torch
INFO:2023-10-30 20:57:45 14:14 init.cpp:121] Registering daemon config loader
WARNING:2023-10-30 20:57:45 14:14 init.cpp:155] function cbapi->getCuptiStatus() failed with error CUPTI_ERROR_MULTIPLE_SUBSCRIBERS_NOT_SUPPORTED (39)
WARNING:2023-10-30 20:57:45 14:14 init.cpp:156] CUPTI initialization failed - CUDA profiler activities will be missing
INFO:2023-10-30 20:57:45 14:14 init.cpp:158] If you see CUPTI_ERROR_INSUFFICIENT_PRIVILEGES, refer to https://developer.nvidia.com/nvidia-development-tools-solutions-err-nvgpuctrperm-cupti
\>>> print(torch.version.cuda)
12.1
\>>> print(torch.__version__)
2.1.0+cu121

# My Test Environment

pytorch: 2.1.0+cu121
cuda toolkit and cuda driver: 12.1

+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 530.30.02              Driver Version: 530.30.02    CUDA Version: 12.1     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                  Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf            Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla V100-SXM2-32GB            Off| 00000000:00:0B.0 Off |                    0 |
| N/A   41C    P0               43W / 300W|      0MiB / 32768MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2-32GB            Off| 00000000:00:0D.0 Off |                    0 |
| N/A   40C    P0               42W / 300W|      0MiB / 32768MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2-32GB            Off| 00000000:00:0E.0 Off |                    0 |
| N/A   41C    P0               43W / 300W|      0MiB / 32768MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   3  Tesla V100-SXM2-32GB            Off| 00000000:00:0F.0 Off |                    0 |
| N/A   40C    P0               42W / 300W|      0MiB / 32768MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   4  Tesla V100-SXM2-32GB            Off| 00000000:00:10.0 Off |                    0 |
| N/A   43C    P0               44W / 300W|      0MiB / 32768MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   5  Tesla V100-SXM2-32GB            Off| 00000000:00:1B.0 Off |                    0 |
| N/A   39C    P0               43W / 300W|      0MiB / 32768MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+

cc @robieta @chaekit @aaronenyeshi @nbcsm @guotuofeng @guyang3532 @gaoteng-git @tiffzhaofb @dzhulgakov @davidberard98

### PR DESCRIPTION
**Pull request by briancoutinho**
Fixes #112389 

## About

PyTorch (Kineto) profiler registers with the profiling daemon Dynolog to enable on-demand profiling. The user should only need to set the env variable `KINETO_USE_DAEMON`. To enable this we need to initialize kineto library early rather than lazily on a PyTorch profiler call. This initialization happens in a static initializer.
- Kineto init function basically registers a callback using the CUDA CUPTI library https://github.com/pytorch/kineto/blob/main/libkineto/src/init.cpp#L130-L148
- However, the above needs the dynamic linking to libcupti.so to have taken place.
- I understand now that static initializations of compilation units will be called before the dynamic linking leading to a segfault in #112389 

image

## Workaround
We add a delay in the initialization that can be configured using the env variable 'KINETO_DAEMON_INIT_DELAY_S'. May not be the best but it could help resolve the issue.

## Testing
Tested this out with [linear_model_example.py](https://github.com/facebookincubator/dynolog/blob/main/scripts/pytorch/linear_model_example.py)
First export the daemon env variable

### Without any delay
```
>$ python3 linear_model_example.py

INFO:2024-02-21 19:34:50 2366287:2366287 init.cpp:131] Registering daemon config loader, cpuOnly =  1
INFO:2024-02-21 19:34:50 2366287:2366287 DaemonConfigLoader.cpp:63] Setting communication fabric enabled = 1
INFO:2024-02-21 19:34:50 2366287:2366287 IpcFabricConfigClient.cpp:93] Setting up IPC Fabric at endpoint: dynoconfigclientb8f91363-d8d6-47a7-9103-197661e28397 status = initialized
INFO:2024-02-21 19:34:50 2366287:2366287 DaemonConfigLoader.cpp:63] Setting communication fabric enabled = 1
INFO:2024-02-21 19:34:50 2366287:2366287 DaemonConfigLoader.cpp:63] Setting communication fabric enabled = 1
cpu
99 1385.468505859375
```

### With 5 seconds delay
```
>$ KINETO_DAEMON_INIT_DELAY_S=5 python3 linear_model_example.py

cpu
99 284.82305908203125
10099 8.817167282104492
INFO:2024-02-21 19:34:26 2359155:2359214 init.cpp:131] Registering daemon config loader, cpuOnly =  1
ERROR: External init callback must run in same thread as registerClient (1782580992 != -1922169024)
INFO:2024-02-21 19:34:26 2359155:2359214 DaemonConfigLoader.cpp:63] Setting communication fabric enabled = 1
INFO:2024-02-21 19:34:26 2359155:2359214 IpcFabricConfigClient.cpp:93] Setting up IPC Fabric at endpoint: dynoconfigclient49270a3f-e913-4ea6-b9e0-cc90a853a869 status = initialized
INFO:2024-02-21 19:34:26 2359155:2359214 DaemonConfigLoader.cpp:63] Setting communication fabric enabled = 1
INFO:2024-02-21 19:34:26 2359155:2359214 DaemonConfigLoader.cpp:63] Setting communication fabric enabled = 1
20099 8.817167282104492
```

### With an invalid delay
```
>$ KINETO_DAEMON_INIT_DELAY_S=abc python3 linear_model_example.py

INFO:2024-02-21 19:35:02 2369647:2369647 init.cpp:131] Registering daemon config loader, cpuOnly =  1
INFO:2024-02-21 19:35:02 2369647:2369647 DaemonConfigLoader.cpp:63] Setting communication fabric enabled = 1
INFO:2024-02-21 19:35:02 2369647:2369647 IpcFabricConfigClient.cpp:93] Setting up IPC Fabric at endpoint: dynoconfigclient0e12a349-af7b-4322-901d-1ff22f91fd4c status = initialized
INFO:2024-02-21 19:35:02 2369647:2369647 DaemonConfigLoader.cpp:63] Setting communication fabric enabled = 1
INFO:2024-02-21 19:35:02 2369647:2369647 DaemonConfigLoader.cpp:63] Setting communication fabric enabled = 1
cpu
```

## Impact
This should not impact any general user. The initialization only occurs if `KINETO_USE_DAEMON` is set in the environment.

### CODE REVIEW

**Comment by anupambhatnagar** [comment]
Looks like an innocuous change since it is guarded by an environment variable. A couple of questions: 

1. Will the trace show function call "get_init_delay" when the env var is set to True?
2. Have you tested this on any example model?

**Review by anupambhatnagar — APPROVED** [review]
LGTM!

**Comment by briancoutinho** [comment]
> Looks like an innocuous change since it is guarded by an environment variable. A couple of questions:
> 
> 1. Will the trace show function call "get_init_delay" when the env var is set to True?
It should not show get_init_delay() as its just called at init time. Not sure if i got your question, let me know.

> 2. Have you tested this on any example model?
Ah yes just updated testplan

File: test/profiler/test_profiler.py [hunk 1]
```diff
@@ -1661,6 +1661,7 @@ def test_nested_tensor_with_shapes(self):
                 self.assertTrue(len(e.input_shapes[0]) > 0)
 
     @patch.dict(os.environ, {"KINETO_USE_DAEMON": "1"})
+    @patch.dict(os.environ, {"KINETO_DAEMON_INIT_DELAY_S": "1"})
```
Comment (aaronenyeshi) [inline test/profiler/test_profiler.py:1664]: May be useful to include the unit, like `KINETO_DAEMON_INIT_DELAY_SEC`

File: torch/csrc/profiler/kineto_client_interface.cpp [hunk 2]
```diff
@@ -73,18 +75,43 @@ class LibKinetoClient : public libkineto::ClientInterface {
...
+    };
 
-    libkineto::api().registerClient(&client);
+    if (std::getenv("KINETO_USE_DAEMON") != nullptr) {
+      int init_delay_s = get_init_delay();
+      if (init_delay_s > 0) {
+        std::thread t([init_delay_s, kineto_init]() {
+          std::this_thread::sleep_for(std::chrono::seconds(init_delay_s));
+          kineto_init();
+        });
```
Comment (aaronenyeshi) [inline torch/csrc/profiler/kineto_client_interface.cpp:109]: Curious what will happen if the user launches the training on a separate "forkserver" process? Will both the parent process and forkserver process perform delayed init? (I've noticed forkserver getting more popular lately).

**Review by aaronenyeshi — APPROVED** [review]
The workaround looks good!

File: torch/csrc/profiler/kineto_client_interface.cpp (hunk 2 above)
Comment (briancoutinho) [reply torch/csrc/profiler/kineto_client_interface.cpp:109]: I hope the env vars get nested :) else both USE_DAEMON and DELAY won't be passed

File: test/profiler/test_profiler.py (hunk 1 above)
Comment (briancoutinho) [reply test/profiler/test_profiler.py:1664]: Ah didn't spell SEC but have the S in the end KINETO_DAEMON_INIT_DELAY_S

**Comment by hi20240217** [comment]
External init callback must run in same thread as registerClient (1744025344 != 318636416)

**Pull request closed without merging**